File: daily/2019-12-19/20 Valid Parenthese.py
class Solution:
    def isValid(self, s):
        # use hash map to solve this question
        # use the stack to store the data
        hash_map = {
            "(": ")",
            "{": "}",
            "[": "]"
        }
        stack = []

        # define a stack
        for i in s:
            if i not in hash_map:
                if stack and hash_map[stack.pop()] == i:
                    continue
                else:
                    return False
            else:
                stack.append(i)

        return stack == []

File: Leetcode First 200 JavaScript solution/20. Valid Parenthese.js
var isValid = function(s) {
    let hashmap = {
        '[' : ']',
        '{' : '}',
        '(' : ')'
    }
    let stack = [];
    for (let char of s) {
        if (char in hashmap) {
            stack.push(char);
            continue;
        } else {
            if (stack.length === 0 || hashmap[stack.pop()] !== char) return false;
        }
    }
    return stack.length === 0;
};

// o(n)
// o(n)
